
## Downsampling populations

All populations were downsampled using a specific number of individuals (n=10). Only 7 populations (out of 40) were not downsampled since they have low sample size (n<10). To downsample each population, we selected individuals 1) without high levels of relatedness (`criterion used to removed highly genetically related individuals`), 2) highest number of genes covered completely with at least 1 read across the genome, and 3) a random selection.

Final downsampled population sample sets are shown in TXT attached file `aedes_aegypti.downsampled_size.txt`.

Next, for each population, we calculate the `genetic diversity (pi)` and `Tajima's D` per nucleotide site at whole genome level. Additionally, we estimate population divergence by using `the Population Branch Statistic (PBS)`  by calculating th FST divergence score between pairwise population comparisons. These analyses were obtained by using the `ANGSD` program.

